src/silicone/utils.py

```diff
@@ -1,12 +1,228 @@
+import numpy as np
+import pandas as pd
+import scipy.interpolate
+
 """
 Utils contains a number of helpful functions that don't belong elsewhere.
 """
 
 
+def find_matching_scenarios(
+    options_df,
+    to_compare_df,
+    variable_follower,
+    variable_leaders,
+    classify_scenarios,
+    classify_models=["*"],
+    return_all_info=False,
+    use_change_not_abs=False,
+):
+    """
+    Groups scenarios and models into different classifications and uses those to
+    work out which group contains a trendline most similar to the data. These
+    combinations may group several models/scenarios together by means of wild cards.
+    Most similar means having the smallest total squared distance between the
+    to_compare_df value of variable_follower and the variable_follower values
+    interpolated in options_df at the variable_leaders points in to_compare_df, i.e.
+    assuming errors only exist in variable_follower.
+    In the event of a tie between different scenario/model classifications, it returns the
+    scenario/model combination that occurs earlier in the input lists, looping through
+    scenarios first.
+
+    Parameters
+    ----------
+    options_df : :obj:`pyam.IamDataFrame`
+        The dataframe containing the data for each scenario and model
+
+    to_compare_df : :obj:`pyam.IamDataFrame`
+        The dataframe we wish to find the scenario group closest to. May contain one
+        or more scenarios, we minimise the least squared errors for all the data
+        colleectively.
+
+    variable_follower : str
+        The variable we want to interpolate and compare to the value in to_compare_df
+
+    variable_leaders : list[str]
+        The variable(s) we want to use to construct the interpolation
+        (e.g. ``["Emissions|CO2"]``). In the event that there are multiple, we
+        interpolate with each one separately and minimise the sum of the squared
+        errors.
+
+    classify_scenarios : list[str]
+        The names of scenarios or groups of scenarios that are possible matches.
+        This may have *s to represent wild cards, hence multiple scenarios will have
+        all their data combined to make the interpolator.
+
+    classify_models : list[str]
+        The names of models or groups of models that are possible matches.
+        This may have *s to represent wild cards, hence multiple models will have
+        all their data combined to make the interpolator.
+
+    return_all_info : bool
+        If True, instead of simply returning the strings specifying the closest
+        scenario/model match, we return all scenario/model combinations in order of
+        preference, along with
+
+    use_change_not_abs : bool
+        If True, the code looks for the trend with the closest *derivatives* rather
+        than the closest absolute value, i.e. closest trend allowing for an offset.
+        This requires data from more than one time.
+
+    Returns
+    -------
+    if return_all_info == False:
```

Comment:
having different return types depending on an input argument is a bit weird, so I'd be tempted to split into two functions (one that always returns all info, one that always returns the tuple of strings) but also fine if you want to just leave it

Reply:
I think I'll leave it unless someone complains, it's directly for human use so we can see what people say